icu4c/source/common/mlbe.cpp

```diff
@@ -128,135 +122,169 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
-        score += (2 * fModel.geti(feature.setTo(u"TW1:", 4)
-                                      .append(elementList[0])
-                                      .append(elementList[1])
-                                      .append(elementList[2])));
+    if (indexList[0] != -1 && indexList[1] != -1 && indexList[2] != -1) {
+        // TW1
+        end = (indexList[3] != -1) ? indexList[3] : codeUts;
+        score += fModel[9].geti(inString.tempSubString(indexList[0], end - indexList[0]));
     }
-    if (elementList[1] != INVALID && elementList[2] != INVALID && elementList[3] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW2:", 4)
-                                      .append(elementList[1])
-                                      .append(elementList[2])
-                                      .append(elementList[3])));
+    if (indexList[1] != -1 && indexList[2] != -1 && indexList[3] != -1) {
+        // TW2
+        end = (indexList[4] != -1) ? indexList[4] : codeUts;
+        score += fModel[10].geti(inString.tempSubString(indexList[1], end - indexList[1]));
     }
-    if (elementList[2] != INVALID && elementList[3] != INVALID && elementList[4] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW3:", 4)
-                                      .append(elementList[2])
-                                      .append(elementList[3])
-                                      .append(elementList[4])));
+    if (indexList[2] != -1 && indexList[3] != -1 && indexList[4] != -1) {
+        // TW3
+        end = (indexList[5] != -1) ? indexList[5] : codeUts;
+        score += fModel[11].geti(inString.tempSubString(indexList[2], end - indexList[2]));
     }
-    if (elementList[3] != INVALID && elementList[4] != INVALID && elementList[5] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW4:", 4)
-                                      .append(elementList[3])
-                                      .append(elementList[4])
-                                      .append(elementList[5])));
+    if (indexList[3] != -1 && indexList[4] != -1 && indexList[5] != -1) {
+        // TW4
+        end = codeUts;
+        score += fModel[12].geti(inString.tempSubString(indexList[3], end - indexList[3]));
     }
     if (score > 0) {
         boundary.addElement(index, status);
         numBreaks++;
     }
 }
 
-int32_t MlBreakEngine::initElementList(const UnicodeString &inString, UChar32* elementList,
-                                         UErrorCode &status) const {
+int32_t MlBreakEngine::initIndexList(const UnicodeString &inString, int32_t* indexList,
+                                     UErrorCode &status) const {
     if (U_FAILURE(status)) {
         return 0;
     }
     int32_t index = 0;
     int32_t length = inString.countChar32();
-    UChar32 w1, w2, w3, w4, w5, w6;
-    w1 = w2 = w3 = w4 = w5 = w6 = INVALID;
+    int32_t i3, i4, i5, i6;
```

Comment:
I don't think we need i3, i4, i5, i6 at all here.

Reply:
Done